### ISSUE TITLE
Embedded DNS server breaks local name resolution in user defined network

### ISSUE THREAD

**Issue by gg7** [docker/docker#31244]
**Description**

On my Docker host I have `search company` in `/etc/resolv.conf`. When using the default bridge network I can do this:

```
george@george:~$ docker run --rm -it dnstester host kafka0
kafka0.company has address 10.0.0.70
```

When using a user-defined network this doesn't work and I'm forced to use FQDNs:

```
george@george:~$ docker network create net1
george@george:~$ docker run --rm -it --net=net1 dnstester host kafka0
Host kafka0 not found: 3(NXDOMAIN)
```

I understand that the embedded DNS server will do service discovery so it will work if there's a "kafka0" container running in the same network, but I still want to resolve unqualified external domain names.

Passing `--dns=x` does not help. I also can't override `ndots:0`:

```
george@george:~$ docker run --rm -it --net=net1 --dns-opt='ndots:1' dnstester cat /etc/resolv.conf | grep ndots
options ndots:1 ndots:0
```

How can I get `kafka0` to resolve when using custom networks?

I consider this a bug because it's breaking networking for me.

**Output of `docker version`:**

```
george@george:~$ docker version
Client:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 21:47:50 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.11.2
 API version:  1.23
 Go version:   go1.5.4
 Git commit:   b9f10c9
 Built:        Wed Jun  1 21:47:50 2016
 OS/Arch:      linux/amd64
```

**Output of `docker info`:**

```
george@george:~$ docker info
Containers: 1
 Running: 1
 Paused: 0
 Stopped: 0
Images: 893
Server Version: 1.11.2
Storage Driver: aufs
 Root Dir: /var/lib/docker/aufs
 Backing Filesystem: zfs
 Dirs: 1237
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins: 
 Volume: local
 Network: bridge null host
Kernel Version: 4.4.0-62-generic
Operating System: Ubuntu 14.04.5 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 62.78 GiB
Name: george
ID: ACSH:EYI5:G36O:2SRA:FWLH:U4GF:VMJM:C4KF:73MH:L273:TCCO:GW3K
Docker Root Dir: /var/lib/docker
Debug mode (client): false
Debug mode (server): false
Username: gg77
Registry: https://index.docker.io/v1/
WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**

Standard 64 bit x86 workstation.

**Comment by thaJeztah** [comment]
ping @sanimej ptal

**Comment by sanimej** [comment]
@gg7 The default behavior with docker DNS is the first query will be the unqualified one (because of `ndots:0`). If that fails, the expectation is that the client will try with the search prefix appended. But this behavior is not consistent. Some glibc implementations take the view that if the resolution fails in the root domain (ie: without any domain name attached) they will give up without trying the `search` domains configure. Here is an example..

With ubuntu based docker image, `host` command fails..
```
vagrant@net-3:~$ docker run -it --rm --net docker --name web --dns-search gmail.com sanimej/ubuntu bash
root@1de7504ec091:/# host smtp
Host smtp not found: 3(NXDOMAIN)
root@1de7504ec091:/# 

root@1de7504ec091:/# ldd --version
ldd (Ubuntu EGLIBC 2.19-0ubuntu6.6) 2.19
Copyright (C) 2014 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
root@1de7504ec091:/# 

root@1de7504ec091:/# cat /etc/resolv.conf 
search gmail.com
nameserver 127.0.0.11
options ndots:0
root@1de7504ec091:/# 
```

With a Centos based image it works ok..
```
vagrant@net-3:~$ docker run -it --rm --net docker --name web --dns-search gmail.com centos bash
[root@83a19c3851c1 /]# host smtp
smtp.gmail.com is an alias for gmail-smtp-msa.l.google.com.
gmail-smtp-msa.l.google.com has address 74.125.28.108
gmail-smtp-msa.l.google.com has address 74.125.28.109
gmail-smtp-msa.l.google.com has IPv6 address 2607:f8b0:400e:c04::6c

[root@83a19c3851c1 /]# ldd --version
ldd (GNU libc) 2.17
Copyright (C) 2012 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
Written by Roland McGrath and Ulrich Drepper.
[root@83a19c3851c1 /]# 

[root@83a19c3851c1 /]# cat /etc/resolv.conf 
search gmail.com
nameserver 127.0.0.11
options ndots:0
[root@83a19c3851c1 /]# 
```

1.13 also has the [fix for handling the user configured `ndots` correctly](https://github.com/docker/libnetwork/pull/1441). You will still see the `ndots:0` in resolv.conf. But the Docker DNS server will not forward the absolute query without the domain name to external servers.

**Comment by sanimej** [comment]
In an ubuntu based container while the `host` fails `getent`  works.  ping works as well. From the docker daemon debugs you can see that `host` command doesn't try with the search domain attached where as ping/getent does.

```
capture for host command

22:48:47.879837 IP ::1.47130 > 127.0.0.11.53280: UDP, length 22
22:48:47.880033 IP 172.22.0.4.57595 > 10.0.2.3.domain: 58350+ A? smtp. (22)
22:48:47.897362 IP 10.0.2.3.domain > 172.22.0.4.57595: 58350 NXDomain 0/1/0 (97)
22:48:47.897541 IP 127.0.0.11.domain > ::1.47130: 58350 NXDomain 0/1/0 (97)

capture for ping

22:48:58.840364 IP ::1.55962 > 127.0.0.11.53280: UDP, length 22
22:48:58.841396 IP 172.22.0.4.36230 > 10.0.2.3.domain: 8100+ A? smtp. (22)
22:48:58.989237 IP 10.0.2.3.domain > 172.22.0.4.36230: 8100 NXDomain 0/1/0 (97)
22:48:58.989416 IP 127.0.0.11.domain > ::1.55962: 8100 NXDomain 0/1/0 (97)
22:48:58.989513 IP ::1.50390 > 127.0.0.11.53280: UDP, length 32
22:48:58.990472 IP 172.22.0.4.43035 > 10.0.2.3.domain: 56521+ A? smtp.gmail.com. (32)
22:48:59.006608 IP 10.0.2.3.domain > 172.22.0.4.43035: 56521 3/0/0 CNAME gmail-smtp-msa.l.google.com., A 74.125.28.109, A 74.125.28.108 (102)
22:48:59.006785 IP 127.0.0.11.domain > ::1.50390: 56521 3/0/0 CNAME gmail-smtp-msa.l.google.com., A 74.125.28.109, A 74.125.28.108 (102)
```
The fix I mentioned in 1.13 is not relevant in cases where the client doesn't try to resolve the name with search domain attached. I am closing this issue since the behavior depends on the client/resolver library implementation and some of them does work ok.

**Comment by gg7** [comment]
@sanimej Thanks for the detailed answer.
